// appex/compiler/Compiler.ts
// Copyright (c) 2013 haydn paterson (sinclair).  All rights reserved.
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//   http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

/// <reference path="../interfaces.ts" />
/// <reference path="../workers/Worker.ts" />
/// <reference path="CompilerResult.ts" />

module appex.compiler {

    /** The appex typescript compiler  */
    export class Compiler implements appex.IDisposable {
        
        private worker: appex.workers.Worker<string, appex.compiler.CompilerResult>;

        constructor() {
            
            this.worker = new appex.workers.Worker<string, appex.compiler.CompilerResult>( this.kernel );
            
        }

        /** compiles a typescript source file. 
        *
        * arguments:
        *       filename - the source file to compile.
        *       callback - a callback for the compilation result.
        */        
        public compile(filename:string, callback:(result:appex.compiler.CompilerResult)=>void) : void {
            
            
            this.worker.call(filename, callback);

        }

        /** disposes the compiler by terminating the worker process. */
        public dispose() : void {

            this.worker.dispose();

        }

        /** the compilation kernel. run in a child_process */
        private kernel (filename:string, callback:(result:appex.compiler.CompilerResult)=>void) : void {
            
            var gather_diagnostics = function(units:typescript.api.Unit[]) : typescript.api.Diagnostic[] {
                
                var diagnostics = [];

                units.forEach((unit) => {
                    
                    unit.diagnostics.forEach((diagnostic) => {

                        diagnostics.push(diagnostic);
                    });
                });

                return diagnostics;
            }
            
            var typescript = <typescript.api>require('typescript.api');

            // typescript.reset(); 

            var errors = [];

            typescript.resolve([filename], (resolved)=> {
                
                if(!typescript.check(resolved)) {

                    callback({
                        
                        javascript  : '',
                        
                        scripts     : null,

                        diagnostics : gather_diagnostics(resolved)
                    });

                    return;
                }

                typescript.compile(resolved, (compiled)=> {
                    
                    if(!typescript.check(compiled)) {

                        callback({
                        
                            javascript  : '',
                        
                            scripts     : null,

                            diagnostics : gather_diagnostics(compiled)
                        });

                        return; 
                    }                   

                    // create compiler result.
                    
                    var content    = [];

                    var scripts    = [];

                    var declaration = [];

                    compiled.forEach((unit) => {
                    
                        content.push      (unit.content);

                        scripts.push      (unit.script);
                    });
                    
                    callback({
                        
                        javascript   : content.join('\n'),

                        scripts      : scripts,

                        diagnostics  : [],
                    });
                });
            });
        }
    }
}
