// appex/compiler/Compiler.ts
/*--------------------------------------------------------------------------

The MIT License (MIT)

Copyright (c) 2013 Haydn Paterson (sinclair) <haydn.developer@gmail.com>

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in
all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
THE SOFTWARE.

---------------------------------------------------------------------------*/

/// <reference path="../interfaces.ts" />
/// <reference path="../workers/Worker.ts" />
/// <reference path="CompilerResult.ts" />

module appex.compiler {

    /** The appex typescript compiler  */
    export class Compiler implements appex.IDisposable {
        
        private worker: appex.workers.Worker<string, appex.compiler.CompilerResult>;

        constructor() {
            
            this.worker = new appex.workers.Worker<string, appex.compiler.CompilerResult>( this.kernel );
            
        }

        /** compiles a typescript source file. 
        *
        * arguments:
        *       filename - the source file to compile.
        *       callback - a callback for the compilation result.
        */        
        public compile(filename:string, callback:(result:appex.compiler.CompilerResult)=>void) : void {
            
            
            this.worker.call(filename, callback);

        }

        /** disposes the compiler by terminating the worker process. */
        public dispose() : void {

            this.worker.dispose();

        }

        /** the compilation kernel. run in a child_process */
        private kernel (filename:string, callback:(result:appex.compiler.CompilerResult)=>void) : void {
            
            var gather_diagnostics = function(units:typescript.api.Unit[]) : typescript.api.Diagnostic[] {
                
                var diagnostics = [];

                units.forEach((unit) => {
                    
                    unit.diagnostics.forEach((diagnostic) => {

                        diagnostics.push(diagnostic);
                    });
                });

                return diagnostics;
            }
            
            var typescript = <typescript.api>require('typescript.api');

            typescript.generateSourceMaps = false;

            var errors = [];

            typescript.resolve([filename], (resolved)=> {
                
                if(!typescript.check(resolved)) {

                    callback({
                        
                        filename    : filename,

                        javascript  : '',
                        
                        scripts     : null,

                        diagnostics : gather_diagnostics(resolved)
                    });

                    return;
                }

                typescript.compile(resolved, (compiled)=> {
                    
                    if(!typescript.check(compiled)) {

                        callback({
                        
                            filename    : filename,

                            javascript  : '',
                        
                            scripts     : null,

                            diagnostics : gather_diagnostics(compiled)
                        });

                        return; 
                    }                   

                    // create compiler result.
                    
                    var content    = [];

                    var scripts    = [];

                    var declaration = [];

                    compiled.forEach((unit) => {
                    
                        content.push      (unit.content);

                        scripts.push      (unit.script);
                    });
                    
                    callback({
                        
                        filename     : filename,

                        javascript   : content.join('\n'),

                        scripts      : scripts,

                        diagnostics  : [],
                    });
                });
            });
        }
    }
}
