// appex/web/app/App.ts
// Copyright (c) 2013 haydn paterson (sinclair).  All rights reserved.
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//   http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

/// <reference path="../../references.ts" />
/// <reference path="../../modules/IModule.ts" />
/// <reference path="../routing/IRouter.ts" />
/// <reference path="IApp.ts" />
/// <reference path="Request.ts" />
/// <reference path="Response.ts" />

module appex.web.app {

    export class App implements appex.web.app.IApp {
        
        public server    : http.Server;

        public request   : appex.web.app.IRequest;

        public response  : appex.web.app.IResponse;

        public next      : Function;

        public module    : appex.modules.IModule;

        public router    : appex.web.routing.IRouter;

        public attribute : any;

        public mime      : appex.web.media.Mime; // should not be here.
        
        constructor(server          : http.Server,
             
                    serverRequest   : http.ServerRequest, 

                    serverResponse  : http.ServerResponse, 

                    next            : Function,

                    module          : appex.modules.IModule,

                    router          : appex.web.routing.IRouter) {
            
            this.server   = server;

            this.request  = new appex.web.app.Request(serverRequest);

            this.response = new appex.web.app.Response(serverResponse);

            this.next     = next;

            this.module   = module;

            this.router   = router;
        }
    }
}
