// appex/web/routing/IndexRoute.ts
/*--------------------------------------------------------------------------

The MIT License (MIT)

Copyright (c) 2013 Haydn Paterson (sinclair) <haydn.developer@gmail.com>

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in
all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
THE SOFTWARE.

---------------------------------------------------------------------------*/

/// <reference path="../../util/TypeCheck.ts" />
/// <reference path="../../modules/IModule.ts" />
/// <reference path="../IContext.ts" />
/// <reference path="IModuleRoute.ts" />

module appex.web.routing {

    /** appex index route. handles functions named index. */
    export class IndexRoute implements appex.web.routing.IModuleRoute {

        /** the pathname used to match this route. */
        public pathname: string;

        /** the type of route this is */
        public type     : string;

        /** arguments:
        *       moduleExport : the module export for this route. (must be a function).
        */
        constructor(public moduleExport: appex.modules.IModuleExport) {

            this.type = 'index';

            this.initialize();
        }

        /** matches the verbs on the attribute (if exists) 
        *  arguments:
        *       context   : the request context
        *       attribute : the attribute for this route.
        */
        private match_attribute_verbs(context: appex.web.IContext, attribute: any): boolean {

            if(attribute.verbs) {

                if(appex.util.TypeCheck.isArray(attribute.verbs)) {

                    for(var n in attribute.verbs) {

                        if(appex.util.TypeCheck.isString(attribute.verbs[n])) {

                            if(attribute.verbs[n].toLowerCase()==context.request.method.toLowerCase()) {

                                return true;
                            }
                        }
                    }
                }
                return false;
            }
            else {

                return true;
            }
        }

        /** matches the urls applied to the attribute (if any)
        *  arguments:
        *       context   : the request context
        *       pathname  : the url pathname to match.
        *       attribute : the attribute for this route.
        */
        private match_attribute_urls(context: appex.web.IContext, pathname: string, attribute: any): boolean {

            if(attribute.urls) {

                if(appex.util.TypeCheck.isArray(attribute.urls)) {

                    for(var n in attribute.urls) {

                        if(appex.util.TypeCheck.isString(attribute.urls[n])) {

                            if(attribute.urls[n] == pathname) {

                                return true;
                            }
                        }
                    }
                }
            }

            return false;
        }


        /** matches a route from the given context 
        *
        *   arguments:
        *       context : the appex context.
        */
        public match(context: appex.web.IContext): boolean {

            var url = node.url.parse(context.request.url);

            var attribute = this.moduleExport.attribute();

            if(this.match_attribute_verbs(context, attribute)) {

                if(this.match_attribute_urls(context, url.pathname, attribute)) {

                    return true;
                }

                return this.pathname == url.pathname;
            }
        }

        /** invokes the target.
        *
        *   arguments:
        *       method  : the method to invoke.
        *       context : the context.
        */
        private invoke(method:any, context: appex.web.IContext): void {

            var stack = [];

            var index = 0;

            var next = function() {

                if(stack.length > 0) {

                    var middleware = stack.pop();

                    context.next = next;

                    middleware(context);
                }
                else {

                    method(context);
                }
            }

            // if use.. invoke the middleware stack.
            if(context.attribute.use) {

                if(appex.util.TypeCheck.isArray(context. attribute.use)) {

                    for(var i= (context.attribute.use.length-1); i >= 0; i--) {

                        if(appex.util.TypeCheck.isFunction(context.attribute.use[i])) {

                            stack.push(context.attribute.use[i])
                        }
                    }

                    next();

                    return;
                }
            }

            method(context);
        }

        /** handles route invocation.
        *
        *   arguments:
        *       app      - the application context.
        *       request  - the nodejs http request.
        *       response - the nodejs http response.
        *       returns  - success if the route was handled.
        */
        public handler(context: appex.web.IContext): boolean {

            var method = this.moduleExport.accessor();

            if(method) {

                context.attribute=this.moduleExport.attribute();

                this.invoke(method,context);

                return true;
            }

            throw new Error('appex.web.routing.ModuleIndexRoute : unable to load accessor from moduleExport.');

            return false;
        }

        /** returns matchable patterns for this route.
        *
        *   arguments:
        *       returns  - an array of urls which match to this route.
        */
        public urls(): string[] {

            return [this.pathname];
        }

        /** initializes this route */
        private initialize(): void {

            var method = <typescript.api.Method>this.moduleExport.type;

            var scope = method.scope;

            if(scope.length > 0) {

                if(scope[0].indexOf('"') !== -1) {

                    scope = scope.slice(1);
                }
            }

            if(scope.length == 0) {

                this.pathname = '/';
            }
            else {

                this.pathname= '/' + scope.join('/');
            }
        }
    }
}
