// appex/web/routing/ModuleJsonRoute.ts
// Copyright (c) haydn paterson (sinclair) 2013.  All rights reserved.
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//   http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

/// <reference path="../../references.ts" />
/// <reference path="../../modules/Module.ts" />
/// <reference path="IRoute.ts" />

module appex.web.routing {

    export class ModuleJsonRoute implements appex.web.routing.IRoute {
        
        public pathname     : string;

        public requestType  : typescript.api.Type;

        public responseType : typescript.api.Type;

        constructor(public moduleExport : appex.modules.ModuleExport) {
            
            this.initialize();
        }

        public match ( request : http.ServerRequest ) : boolean {
        
            var url = node.url.parse(request.url);

            return this.pathname == url.pathname;
        }

        public handler( context : any, request:http.ServerRequest, response:http.ServerResponse, next?:()=>void ) : boolean {
            
            var accessor = this.moduleExport.accessor();

            if(accessor) {

                this.read_request(request, (input) => {

                    accessor( context, input, (output) => {
                                
                        this.write_response(output, response);
                    });
                });

                return true;
            }

            throw new Error("appex.web.routing.ModuleJsonRoute : unable to load accessor from moduleExport.");

            return false;
        }


        private read_request (request:http.ServerRequest, callback:(obj:any) => void) : void {
           
            var buffer = [];

            request.on('data', (chunk) => {  buffer.push(chunk); });

            request.on('end', () => { 
                
                try  {

                    var obj = JSON.parse( buffer.join('') );

                    callback(obj);
                } 
                catch(error) {

                    callback(null);
                }
            }); 
        }

        private write_response (obj:any, response:http.ServerResponse): void {
        
            try {

                var json = JSON.stringify(obj, null, '');
                
                response.writeHead(200, {'content-type' : 'application/json'});

                response.write(json);

                response.end();
            } 
            catch(error) {

                response.writeHead(500, {'content-type' : 'application/json'});

                response.write(error.toString());

                response.end();
            }            
        }
        
        private initialize() : void {
            
            var method = <typescript.api.Method>this.moduleExport.type;

            try 
            {
                this.requestType  = method.parameters[1].type;
            
                this.responseType = method.parameters[2].type.signature.parameters[0].type;
            } 
            catch(e) 
            {
                throw new Error('appex.web.routing.ModuleJsonRoute : unable to load request and response type.')     
            }

            var scope  = method.scope;

            if(scope.length > 0) {

                if(scope[0].indexOf('"') !== -1) {
                    
                    scope = scope.slice(1);
                }
            }

            if(scope.length == 0) 
            {       
                this.pathname = '/' + method.name;
            } 
            else  
            {
                this.pathname = '/' + scope.join('/') + '/' + method.name;
            } 
        }
    }
}
