// appex/web/Context.ts
// Copyright (c) 2013 haydn paterson (sinclair).  All rights reserved.
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//   http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

/// <reference path="../modules/IModule.ts" />
/// <reference path="routing/IRouter.ts" />
/// <reference path="IContext.ts" />

module appex.web {

    export class Context implements appex.web.IContext {
        
        /** the appex request */
        public request   : appex.web.IRequest;

        /** the appex response */
        public response  : appex.web.IResponse;

        /** the appex cascade */
        public cascade   : any;

        /** the next function */
        public next      : Function;

        /** the appex router */
        public router    : appex.web.routing.IRouter;
        
        /** the current executing module */
        public module    : appex.modules.IModule;
        
        /** mime utility */
        public mime      : appex.web.media.Mime;          
    }
}
