// appex/web/Context.ts
/*--------------------------------------------------------------------------

The MIT License (MIT)

Copyright (c) 2013 Haydn Paterson (sinclair) <haydn.developer@gmail.com>

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in
all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
THE SOFTWARE.

---------------------------------------------------------------------------*/

/// <reference path="../modules/IModule.ts" />
/// <reference path="../schema/JsonSchema.ts" />
/// <reference path="routing/IRouter.ts" />
/// <reference path="sitemap/Sitemap.ts" />
/// <reference path="IContext.ts" />

module appex.web {

    export class Context implements appex.web.IContext {
        
        /** the appex request */
        public request     : appex.web.IRequest;

        /** the appex response */
        public response    : appex.web.IResponse;

        /** the appex attribute */
        public attribute   : any;

        /** the next function */
        public next        : Function;

        /** the appex router */
        public router      : appex.web.routing.IRouter;
        
        /** the appex sitemap */
        public sitemap     : appex.web.sitemap.Sitemap;

        /** the current executing module */
        public module      : appex.modules.IModule;

        /** the appex schema api */
        public schema      : appex.schema.JsonSchema;

        /** the appex template engine */
        public template    : appex.templates.Engine;
        
        /** mime utility */
        public mime        : appex.web.media.Mime;          
    }
}
