// appex/web/Cookie.ts
// Copyright (c) 2013 haydn paterson (sinclair).  All rights reserved.
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//   http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

/// <reference path="../references.ts" />

module appex.web {

    export class Cookie {
             
        public maxage   : string;

        public domain   : string;

        public path     : string;

        public expires  : Date;

        public httpOnly : boolean;

        public secure   : boolean;

        public key      : string;

        public value    : string;
        
        constructor() {
            
            // todo: cookie implementation.

        }
    }

    export class CookieCollection {

        public cookies : Cookie [];

        constructor() {
        
            // todo: add cookie api
        }

        public set(name:string, value:string) : void {
            
        }

        public get (name:string) : string {
            
            
            return null;
        }
    }
}
