// appex/web/IRequest.ts
/*--------------------------------------------------------------------------

The MIT License (MIT)

Copyright (c) 2013 Haydn Paterson (sinclair) <haydn.developer@gmail.com>

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in
all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
THE SOFTWARE.

---------------------------------------------------------------------------*/

/// <reference path="../references.ts" />
/// <reference path="Cookies.ts" />

module appex.web {

    /** appex request body api */
    export class RequestBody {
    
        constructor(public request:http.ServerRequest) { }

        /** receives data submitted for this request
        *   as a string. 
        *   arguments:
        *       callback: the callback to receive data on
        */
        public recv(callback: (data:string) => void) : void {
            
            var buffer = []

            this.request.on('data', (data) => { buffer.push(data) })
            
            this.request.on('end',  () => { callback(buffer.join('')) })
        }

        /** receives json data submitted for this request
        *   and returns as a object.
        *   arguments:
        *       callback: the callback to receive data on
        */
        public json(callback: (obj:any) => void) : void {
            
            this.recv((data:string) => {

                try {

                    var obj = JSON.parse(data)

                    callback(obj)
                }
                catch(e) {

                    callback(null)
                }
            })
        }

        /** receives form submitted data for this request
        *   and returns as a object.
        *   arguments:
        *       callback: the callback to receive data on
        */
        public form(callback: (obj:any) => void) : void {
            
            this.recv((data:string) => {

                try {

                    var obj = node.querystring.parse(data)

                    callback(obj)
                }
                catch(e) {

                    callback(null)
                }
            })
        }
    }

    /** A appex http request */
    export interface IRequest extends http.ServerRequest {

        /** headers */
        headers : any;
        
        /** http request body */
        body    : appex.web.RequestBody;

        /** http cookie collection. */
        cookies : appex.web.CookieCollection;
    }

    /** patches the nodejs http request with appex request api. 
    *   
    *   arguments:
    *       request  - the nodejs http request
    *       returns  - nodejs http + appex request api.
    */
    export function BindRequest(request: http.ServerRequest) : appex.web.IRequest {
        
        var _request:any = <appex.web.IRequest>request;

        if(!_request.body)     _request.body    = new appex.web.RequestBody(request);

        if(!_request.cookies)  _request.cookies = new appex.web.CookieCollection(request);
        
        return _request;
    }
}
