// appex/util/TypeCheck.ts
/*--------------------------------------------------------------------------

The MIT License (MIT)

Copyright (c) 2013 Haydn Paterson (sinclair) <haydn.developer@gmail.com>

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in
all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
THE SOFTWARE.

---------------------------------------------------------------------------*/

module appex.util {

    export class TypeCheck {
        
        /** test to see if this object is a boolean */
        public static isBoolean(object:any) : boolean {

            return (typeof object) === 'boolean';
        }


        /** test to see if this object is a integer */
        public static isInteger(object:any) : boolean {

            return (typeof object === "number") && Math.floor(object) === object;
        }

        /** test to see if this object is a number */
        public static isNumber(object:any) : boolean {

            return (typeof object === "number");
        }

        /** test to see if this object is an array */
        public static isArray(object:any) : boolean {

            return Object.prototype.toString.call( object ) === '[object Array]';
        }

        /** tests to see if this object is a string */
        public static isString(object:any) : boolean {

            return (typeof object) === 'string';
        }

        /** tests to see if this object is a object */
        public static isObject(object:any) : boolean {
            
            return (typeof object) === 'object' && !TypeCheck.isArray(object);
        }

        /** tests to see if this object is a function */
        public static isFunction(object:any) : boolean {
        
            return (typeof object) === 'function';
        }
    }
}
