// appex/workers/kernel.js
// appex - Copyright (c) sinclair 2013.  All rights reserved.
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//   http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

var vm       = require('vm');

var delegate = null;

// initialization of kernel.

function setup(message) {
    
    process.mainModule.filename = message.filename;
    
    var sandbox = {	

        __filename  : __filename,

        __dirname   : __dirname,	    

        global	    : global,

        process     : process,

        require     : require,	

        console     : console,

        exports     : null,   
    };
	
    var script = vm.createScript('exports = ' + message.body.code, 'worker.js' );

    script.runInNewContext ( sandbox );
	
    delegate = sandbox.exports;
}

// invocation of delgate.
function call(message) {
    
    delegate(message.body, function(response) {

        process.send({

            messageid : message.messageid,

            type      : 'call',

            body      : response
        });
    });
}

// event handling.
process.on('message', function(message) {

    switch(message.type) {

        case "setup":

            setup(message);

            break;

        case "call":

            call(message);

            break;
    }    
});
