// appex/servicehost/Meta.ts
/// <reference path="../activation/ActivationContext.ts" />

module appex.servicehost {
    
    export class Type {
        
        constructor() {
        
            
        }
    }
    
    export class Method {
    
        public name  : string;

        public input :
    }

    export class Service {
    
        public methods:Method[];

        constructor() {
            
            this.methods = [];

        }

    }
    
    export class Meta {
       
        public activation : appex.activation.ActivationContext;

        public types      : Type[];

        public services   : Service[];
        
        constructor(activation:appex.activation.ActivationContext) {

            this.activation = activation;

            //console.log(this.activation.runtime.reflection.types);

            this.types      = [];

            this.services   = [];

            for(var n in activation.handles) {
                
                //console.log(activation.handles[n].typename);
            }   
        }
    }
}
