// appex/servicehost/Route.ts
// Copyright (c) sinclair 2013.  All rights reserved.
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//   http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

/// <reference path="../decl/references.ts" />

module appex.servicehost {

    export class Route {
        
        public path     : string;   // the path to this route.

        public async    : boolean;  // the calling convention.

        public instance : any;      // the instance this route refers.

        public method   : string;   // the method in which to call.
        
        public input    : typescript.api.Type; // the input type.

        public output   : typescript.api.Type; // the output type.

        public call(input:any, callback:(output:any) => void) : any
        {
            console.log('calling -> ' + this.method)

            return this.instance[this.method](input, callback);
        }
    }
}
