// appex/http/RuntimeOptions.ts
/// <reference path="../references.ts" />

module appex {

    export interface IRuntimeOptions {
        
        sourcefile      : string;

        devmode ?       : boolean;

        logging ?       : boolean; 

        stdout  ?       : WritableStream;

        stderr  ?       : WritableStream;
    }

    export class RuntimeOptions implements appex.IRuntimeOptions {

        public sourcefile      : string;

        public devmode         : boolean;

        public logging         : boolean; 

        public stdout          : WritableStream;

        public stderr          : WritableStream; 
        
        constructor(options:IRuntimeOptions) {
            
            this.sourcefile = null;

            this.devmode    = false;

            this.logging    = false;

            this.stdout     =  process.stdout;

            this.stdout     =  process.stdout;

            if(options.sourcefile) this.sourcefile = options.sourcefile;

            if(options.devmode)    this.devmode    = options.devmode;

            if(options.logging)    this.logging    = options.logging;

            if(options.stdout)     this.stdout     = options.stdout;

            if(options.stderr)     this.stderr     = options.stderr;  

            if(this.sourcefile == null) throw new Error('appex: sourcefile required.');   
        }
    }
}
