// appex/modules/IModule.ts
// Copyright (c) 2013 haydn paterson (sinclair).  All rights reserved.
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//   http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

/// <reference path="../references.ts" />

module appex.modules {

    /** interface for module exports */
    export interface IModuleExport  {
        
        type        : typescript.api.ReflectedType;

        accessor()  : any;

        attribute() : any;
    }
    
    /** interface for modules */
    export interface IModule extends appex.IDisposable {
    
        exports : appex.modules.IModuleExport[];

        /** the vm context. (exported objects accessible on context.exports) */
        context     : any;

        /** the javascript this module is mapping */
        javascript  : string;

        /** type reflection meta data for this module */
        reflection  : appex.modules.Reflection;

        /** compiler diagnostics for this module. */
        diagnostics : typescript.api.Diagnostic[];
    }
}
