// appex/modules/Reflection.ts
/*--------------------------------------------------------------------------

The MIT License (MIT)

Copyright (c) 2013 Haydn Paterson (sinclair) <haydn.developer@gmail.com>

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in
all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
THE SOFTWARE.

---------------------------------------------------------------------------*/

/// <reference path="../references.ts" />

module appex.modules {
    
    /** json schema object. */
    export class SchemaObject {

        public id         : string;
        
        public type       : string;
        
        public properties : any;

        public items      : {};

        public required   : string[];
    }

    /** module reflection api */
    export class Reflection {
        
        /** arguments:
        *
        *      scripts: the scripts return from a appex compile.
        */
        constructor(public scripts: typescript.api.Script[]) { }


        /** returns a json schema 
        *
        *   arguments:
        *       qualifier - the fully qualified name of the type.
        */
        public schema(qualifier:string) : SchemaObject {

            var that = this;

            //----------------------------------------
            // fully qualify a type name.
            //----------------------------------------
            var _qualified_name = function(reflected_type:typescript.api.ReflectedType) : string {
            
                var stack = [];

                for(var n in reflected_type.scope) {
                
                    if(reflected_type.scope[n].indexOf('\"') == -1) {
                        
                        stack.push(reflected_type.scope[n])
                    }
                }
                stack.push(reflected_type.name);

                return stack.join('.');
            }

            //----------------------------------------
            // load schema object.
            //----------------------------------------
            var _load = function(reflected_type:typescript.api.ReflectedType) : SchemaObject {

                //----------------------------------------
                // loads base class properties.
                //----------------------------------------
                var _load_class_variables = function(_class:typescript.api.Class) : typescript.api.Variable [] {

                    var variables = [];

                    for(var n in _class.extends) {

                        var _base_class = <typescript.api.Class>that.get( _qualified_name( _class.extends[n] ) );
                    
                        if(_base_class) {
                        
                            var _variables = _load_class_variables(_base_class);

                            for(var m in _variables) {
                        
                                variables.push(_variables[m]);
                            }              
                        }
                    }


                    for(var n in _class.variables) {
                           
                        var variable = <typescript.api.Variable>_class.variables[n];

                        variables.push(variable);
                    }

                    return variables;
                }

                //----------------------------------------
                // class or interface.
                //----------------------------------------

                if(reflected_type.identifier == 'class' || reflected_type.identifier == 'interface') {

                    var _class      = <typescript.api.Class>reflected_type;

                    var type        = new SchemaObject();

                    type.id         = _class.name;
                
                    type.type       = 'object';

                    // load up the variables.

                    var variables = _load_class_variables(_class);

                    if(variables.length > 0) {
                
                        type.properties = {};
                    }

                    // add them to this object.

                    for(var i = 0; i < variables.length; i += 1) {

                        var variable = <typescript.api.Variable>variables[i];
                        
                        if(variable.isPublic || _class.identifier == 'interface') {
                        
                            type.properties[variable.name] = _load(variable);

                            if(!variable.isOptional) {
                                
                                type.required = type.required || [];

                                type.required.push(variable.name);
                            }
                        }
                    }

                    return type;
                }

                //----------------------------------------
                // variable
                //----------------------------------------
                if(reflected_type.identifier == 'variable') 
                {
                    var variable        = <typescript.api.Variable>reflected_type;

                    var _reflected_type = that.get( _qualified_name( variable.type ) );

                    if(_reflected_type) 
                    {
                        if(variable.type.arrayCount == 0) 
                        {
                            return _load(_reflected_type);
                        }
                        else
                        {
                            var array_type   = new SchemaObject();

                            array_type.id    = variable.name;

                            array_type.type  = 'array';

                            array_type.items = {};

                            array_type.items['type'] = _load(_reflected_type);

                            return array_type;
                        }
                    }
                    else
                    {
                        if(variable.type.arrayCount == 0) 
                        {
                            var _single_type = new SchemaObject();

                            _single_type.id    = variable.name;

                            _single_type.type  = variable.type.name;
                        
                            return _single_type;                        
                        }
                        else 
                        {    
                            var _array_type = new SchemaObject();

                            _array_type.type  = 'array';

                            _array_type.items = {};

                            _array_type.items['type'] = variable.type.name;

                            return _array_type;
                        }
                    }
                }

                return null;
            };

            var reflected_type = this.get(qualifier);

            if(reflected_type) {

                return _load(reflected_type);
            }

            return null;
        }


        /** returns a reflected type.
        *
        *   arguments:
        *       qualifier - the fully qualified name of the type.
        */
        public get(qualifier:string) : typescript.api.ReflectedType {
            
            var stack     = [];

            for(var n in this.scripts) {

                var reflected_type = this.scan_script( qualifier, stack, this.scripts[n]);

                if(reflected_type) {
                
                    return reflected_type;
                }
            }

            return null;
        }

        private scan_variable  (qualifier:string, stack:string[], variable:typescript.api.Variable) : typescript.api.ReflectedType {
            
            stack.push(variable.name);

            if(qualifier == stack.join('.')){
            
                return variable;
            }

            stack.pop();

            return null;
        }

        private scan_method    (qualifier:string, stack:string[], method:typescript.api.Method) : typescript.api.ReflectedType {
            
            stack.push(method.name);

            if(qualifier == stack.join('.') ){

                return method;
            }

            stack.pop();

            return null;
        }
        
        private scan_class     (qualifier:string, stack:string[], _class:typescript.api.Class) : typescript.api.ReflectedType {

            stack.push(_class.name)

            if(qualifier == stack.join('.')){
            
                return _class;
            }

            for(var n in _class.methods) {

                var reflected_type = this.scan_method( qualifier, stack, _class.methods[n] );

                if(reflected_type) {
                
                    return reflected_type;
                }
            }

            for(var n in _class.variables) {

                var reflected_type = this.scan_variable( qualifier, stack, _class.variables[n] );

                if(reflected_type) {
                
                    return reflected_type;
                }
            }

            stack.pop();

            return null;
        }

        private scan_interface ( qualifier:string, stack:string[], _interface:typescript.api.Interface) : typescript.api.ReflectedType {

            stack.push(_interface.name);

            if(qualifier == stack.join('.')){
            
                return _interface;
            }

            for(var n in _interface.methods) {

                var reflected_type = this.scan_method( qualifier, stack, _interface.methods[n] );

                if(reflected_type) {
                
                    return reflected_type;
                }
            }

            for(var n in _interface.variables) {

                var reflected_type = this.scan_variable( qualifier, stack, _interface.variables[n] );

                if(reflected_type) {
                
                    return reflected_type;
                }
            }

            stack.pop();

            return null;
            
        }

        private scan_import    ( qualifier:string, stack:string[], _import:typescript.api.Import) : typescript.api.ReflectedType {

            stack.push(_import.name);

            if(qualifier == stack.join('.')){
            
                return <any>_import;
            }  
            
            stack.pop();

            return null; 
        }

        private scan_module    ( qualifier:string, stack:string[], _module:typescript.api.Module) : typescript.api.ReflectedType {
            
            if(_module.name.indexOf('\"') == -1) {

                stack.push(_module.name); 
            }

            if(qualifier == stack.join('.')){
            
                return _module;
            }

            for(var n in _module.modules) {

                var reflected_type = this.scan_module( qualifier, stack, _module.modules[n] );

                if(reflected_type) {
                
                    return reflected_type;
                }
            }

            for(var n in _module.imports) {

                var reflected_type = this.scan_import( qualifier, stack, _module.imports[n] );

                if(reflected_type) {
                
                    return reflected_type;
                }
            }

            for(var n in _module.interfaces) {

                var reflected_type = this.scan_interface( qualifier, stack, _module.interfaces[n] );

                if(reflected_type) {
                
                    return reflected_type;
                }
            } 
                  
            for(var n in _module.classes) {

                var reflected_type = this.scan_class( qualifier, stack, _module.classes[n] );

                if(reflected_type) {
                
                    return reflected_type;
                }
            }

            for(var n in _module.methods) {

                var reflected_type = this.scan_method( qualifier, stack, _module.methods[n] );

                if(reflected_type) {
                
                    return reflected_type;
                }
            }

            for(var n in _module.variables) {

                var reflected_type = this.scan_variable( qualifier, stack, _module.variables[n] );

                if(reflected_type) {
                
                    return reflected_type;
                }
            }

            if(_module.name.indexOf('\"') == -1) {

                stack.pop();
            }

            return null;
        }

        private scan_script    (qualifier:string, stack:string[], script:typescript.api.Script) : typescript.api.ReflectedType {
            
            for(var n in script.modules) {

                var reflected_type = this.scan_module( qualifier, stack, script.modules[n] );

                if(reflected_type) {
                
                    return reflected_type;
                }
            }

            for(var n in script.interfaces) {

                var reflected_type = this.scan_interface( qualifier, stack, script.interfaces[n] );

                if(reflected_type) {
                
                    return reflected_type;
                }
            } 
                  
            for(var n in script.classes) {

                var reflected_type = this.scan_class( qualifier, stack, script.classes[n] );

                if(reflected_type) {
                
                    return reflected_type;
                }
            }

            for(var n in script.methods) {

                var reflected_type = this.scan_method( qualifier, stack, script.methods[n] );

                if(reflected_type) {
                
                    return reflected_type;
                }
            }

            for(var n in script.variables) {

                var reflected_type = this.scan_variable( qualifier, stack, script.variables[n] );

                if(reflected_type) {
                
                    return reflected_type;
                }
            }

            return null;
        } 
    }
}
